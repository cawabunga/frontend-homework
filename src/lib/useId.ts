import { useMemo } from 'react'

export const useId = (): string => {
  return useMemo(() => `id-${Math.random()}`, [])
}
