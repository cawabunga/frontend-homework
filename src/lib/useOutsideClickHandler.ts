import { useEffect } from 'react'

export const useOutsideClickHandler = (
  targetEl: Element | null,
  handler: (event: Event) => void,
  capture: boolean = false
) => {
  useEffect(() => {
    if (!targetEl) return

    const listener = (event: Event) => {
      if (!(event.target instanceof Node && targetEl.contains(event.target))) {
        handler(event)
      }
    }

    document.addEventListener('click', listener, capture)

    return () => {
      document.removeEventListener('click', listener, capture)
    }
  }, [capture, handler, targetEl])
}
