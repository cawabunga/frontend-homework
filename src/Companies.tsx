import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { ReduxState, Company } from './types'
import { getCompanies, getSelectedCompanyId } from './selectors'
import { setSelectedCompanyId } from './actions'

import CompanyLink from './CompanyLink'
import { MenuList } from './menu/MenuList'
import css from './Companies.module.scss'
import { useId } from './lib/useId'

type ReduxProps = {
  companies: Array<Company>
  selectedCompanyId: Company['id'] | null
}

type DispatchProps = {
  setSelectedCompanyId: (companyId: Company['id']) => void
}

export const Companies = ({
  companies,
  selectedCompanyId,
  setSelectedCompanyId,
}: ReduxProps & DispatchProps) => {
  const id = useId()

  return (
    <>
      <div className={css.heading} aria-owns={id}>
        Your companies
      </div>

      <MenuList spacing={0} id={id}>
        {companies.map((company) => (
          <CompanyLink
            key={company.id}
            company={company}
            onClick={() => setSelectedCompanyId(company.id)}
            isActive={company.id === selectedCompanyId}
          />
        ))}
      </MenuList>
    </>
  )
}

export default connect(
  createStructuredSelector<ReduxState, ReduxProps>({
    companies: getCompanies,
    selectedCompanyId: getSelectedCompanyId,
  }),
  { setSelectedCompanyId }
)(Companies)
