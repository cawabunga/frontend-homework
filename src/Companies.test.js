import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { shallow } from 'enzyme'
import { Companies } from './Companies'
import { useId } from './lib/useId'

jest.mock('./lib/useId')

describe('<Companies />', () => {
  const companies = [
    { id: 1, name: 'Dummy company' },
    { id: 2, name: 'Smarty company' },
  ]

  it('renders list of company links', () => {
    useId.mockReturnValue('id')
    expect(shallow(<Companies companies={companies} />)).toMatchSnapshot()
  })

  it('calls setSelectedCompanyId() function on click', () => {
    const mock = jest.fn()
    const { getByText } = render(
      <Companies
        companies={companies}
        selectedCompanyId={null}
        setSelectedCompanyId={mock}
      />
    )

    expect(mock).not.toHaveBeenCalled()
    userEvent.click(getByText('Smarty company'))
    expect(mock).toHaveBeenCalledWith(2)
  })
})
