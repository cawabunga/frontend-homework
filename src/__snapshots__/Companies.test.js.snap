// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Companies /> renders list of company links 1`] = `
<Fragment>
  <div
    aria-owns="id"
    className="heading"
  >
    Your companies
  </div>
  <MenuList
    id="id"
    spacing={0}
  >
    <CompanyLink
      company={
        Object {
          "id": 1,
          "name": "Dummy company",
        }
      }
      isActive={false}
      key="1"
      onClick={[Function]}
    />
    <CompanyLink
      company={
        Object {
          "id": 2,
          "name": "Smarty company",
        }
      }
      isActive={false}
      key="2"
      onClick={[Function]}
    />
  </MenuList>
</Fragment>
`;
